File: dump.py
#!/usr/bin/env python3
import argparse
import os
import subprocess
import sys

def main():
    parser = argparse.ArgumentParser(
        description="Disassemble a binary using llvm-objdump and view output via less.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "file",
        nargs="?",
        default="build/kern/untitled_bare_metal",
        help="Target file for llvm-objdump"
    )
    parser.add_argument(
        "--format",
        default="att",
        choices=["att", "intel"],
        help="Assembly format"
    )
    parser.add_argument(
        "-f", "--functions",
        nargs="+",
        type=str,
        help="Filter functions to disassemble"
    )
    args = parser.parse_args()

    target_file = args.file
    assembly_format = args.format

    if not os.path.isfile(target_file):
        print(f"Error: {target_file} does not exist", file=sys.stderr)
        print(f"Usage: {sys.argv[0]} [file]", file=sys.stderr)
        sys.exit(1)

    objdump_binary = "llvm-objdump"

    # Set intel_option if the format is intel
    intel_option = ""
    if assembly_format == "intel":
        intel_option = "-Mintel"

    # Build the command
    cmd = [objdump_binary, "-d", "-S"]
    if intel_option:
        cmd.append(intel_option)
    cmd.extend(["-j.text", "-j.text.low", "-j.init", "-j.fini", target_file])

    # Filter functions
    if args.functions:
        cmd.append("--disassemble-symbols=" + ",".join(args.functions))

    # Execute the command and pipe the output to less
    try:
        # Launch llvm-objdump; its stdout is piped to less.
        p1 = subprocess.Popen(cmd, stdout=subprocess.PIPE)
        p2 = subprocess.Popen(["less"], stdin=p1.stdout)
        # Close p1.stdout in the parent so that p1 receives a SIGPIPE if less exits.
        p1.stdout.close()
        p2.communicate()
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception as e:
        print(f"Error executing command: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: dump.rb
#!/usr/bin/env ruby
# Disassembly viewer - shows disassembly with source interleaved

require 'optparse'

# Default configuration
options = {
  file: ARGV[0] || 'build/kern/untitled_bare_metal',
  objdump: 'llvm-objdump',
  format: 'intel'
}

parser = OptionParser.new do |opts|
  opts.banner = "Usage: #{$0} [file] [options]"
  opts.separator ""
  opts.separator "Disassemble kernel binary with source interleaving"
  opts.separator ""

  opts.on("-f", "--format FORMAT", [:intel, :att],
          "Assembly format (intel, att; default: intel)") do |f|
    options[:format] = f
  end

  opts.on("--objdump PATH", "Path to objdump binary (default: llvm-objdump)") do |v|
    options[:objdump] = v
  end

  opts.on("-h", "--help", "Show this help message") do
    puts opts
    exit
  end
end

parser.parse!

# Update file from remaining args if provided
options[:file] = ARGV[0] if ARGV[0]

# Check if binary exists
unless File.file?(options[:file])
  warn "Error: #{options[:file]} does not exist"
  warn "Usage: #{$0} [file]"
  exit 1
end

# Build objdump command
cmd = [
  options[:objdump],
  '-d',           # Disassemble
  '-S',           # Interleave source
]

# Add Intel syntax option if requested
cmd << '-Mintel' if options[:format] == :intel

# Specify sections to disassemble
cmd += [
  '-j.text',
  '-j.text.low',
  '-j.init',
  '-j.fini',
  options[:file]
]

# Pipe to less for pagination
pipe_cmd = "#{cmd.map { |arg| arg.include?(' ') ? "'#{arg}'" : arg }.join(' ')} | less"
exec(pipe_cmd)
